### PR TITLE
les, p2p/discv5: implement in-protocol payment

### PR DESCRIPTION
**Pull request by zsfelfoldi**
This is a WIP PR for the in-protocol payment logic. Currently implemented features:
- a discv5 extension for protocol negotiation ("talkRequest/talkResponse") and a dummy handler for the "les" domain
- a new clientPool function for calculating the necessary additional amount of token balance in order to fulfill a requested connection/capacity update (the core logic is also deduplicated)

Coming soon:
- new LES/4 and "LESTALK" UDP messages necessary for payment processing and price negotiation
- glue logic for the token sale and payment receiver modules

### CODE REVIEW

File: les/balance.go [hunk 1]
```diff
@@ -101,6 +104,31 @@ func (bt *balanceTracker) balanceToPriority(b balance) int64 {
 	return int64(b.neg)
 }
 
+func (bt *balanceTracker) posBalanceMissing(targetPriority int64, targetCapacity uint64, after time.Duration) uint64 {
```
Comment (rjl493456442) [inline les/balance.go]: Please add the function description and relative unit tests.

Comment (rjl493456442) [reply les/balance.go]:
```go
func (bt *balanceTracker) budget(targetPriority int64, targetCapacity uint64, after time.Duration)  uint64 {
	current := bt.balanceToPriority(bt.balance)
	if targetPriority >= current {
		if bt.negTimeFactor == 0 {
			return 0
		}
		diff :=  time.Duration(float64(targetPriority - current)/bt.negTimeFactor)
		if diff >= after {
			return 0
		}
		after -= diff
	}
	// We have to cover the connection expense by positive balance.
	if targetPriority >= 0 {
		targetPriority = -1
	}
	budget := uint64(float64(^targetPriority)*float64(targetCapacity) + float64(after)*bt.timeFactor)
	if budget >= maxBalance {
		return math.MaxUint64 // target not reachable
	}
	if budget > bt.balance.pos {
		return budget - bt.balance.pos
	}
	return 0
}
```

Comment (rjl493456442) [reply les/balance.go]: Just open a question here. So we can only guarantee the idle connection by the budget. However request can also cost a lot. TBH I think request cost should be much higher than connection cost.

So should we consider the estimated request cost when we calculate the budget? Although we can never have a correct number.

File: les/server_handler.go [hunk 2]
```diff
@@ -948,3 +950,7 @@ func (h *serverHandler) broadcastHeaders() {
 		}
 	}
 }
+
+func (h *serverHandler) talkRequestHandler(id enode.ID, addr *net.UDPAddr, payload []byte) ([]byte, bool) {
```
Comment (rjl493456442) [inline les/server_handler.go]: Does it mean servers can only receive the UDP packet and then reply? 
So in the following scenarios:

* Client should ask the current token price
* Client should ask the current balance to determine whether they need to buy more tokens for stable connection

**Comment by zsfelfoldi** [comment]
Superseded by https://github.com/ethereum/go-ethereum/pull/20517

**Pull request closed without merging**